File: Java/Airdata/src/main/java/com/plessmann/airdata/datasource/Countries.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.plessmann.airdata.datasource;

import com.plessmann.airdata.models.Country;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 *
 * @author Plesse
 */
public class Countries {
    
    protected static Countries me = null;
    
    private final List<Country> countries;

    public Countries() {
        countries = new ArrayList<>();
    }
    
    public static synchronized Countries getInstance() {
        if(me == null){
            me = new Countries();
        }
        return me;
    }
    
    public void addCountry(Country country) {
        countries.add(country);
    }
    
    public List<Country> getCountryByName(String name) {
        return countries.stream().filter(country -> country.getName().toLowerCase().startsWith(name) || country.getCode().toLowerCase().equals(name)).collect(Collectors.toList());
    }
    
    public List<Country> sortByAirportsQty(boolean top, boolean calc, int maxElements){
        if(calc){
            countries.forEach((country) -> country.setAirportsQty(Airports.getInstance().qtyByCountry(country)));
        }
        countries.sort((p1, p2) -> top ? p2.getAirportsQty() - p1.getAirportsQty() : p1.getAirportsQty() - p2.getAirportsQty());
        return countries.subList(0, maxElements);
    }
    
    public List<Country> sortByName() {
        countries.sort((p1, p2) -> p1.getName().compareTo(p2.getName()));
        return countries;
    }

    public int count() {
        return countries.size();
    }

    public void clear() {
        countries.clear();
    }
}

File: Scala/Airdata/src/com/plessmann/airdata/datasource/Countries.scala
package com.plessmann.airdata.datasource

import com.plessmann.airdata.models.Country

/**
 * Created by Plesse on 12/8/2016.
 */
object Countries {

  var countries: scala.collection.mutable.MutableList[Country] = scala.collection.mutable.MutableList();

  def add(country: Country): Unit ={
    countries.+=(country);
  }

  def count() ={
    countries.size;
  }

  def getCountryByName(name: String) ={
    countries.filter((x) => x.name.value.toLowerCase.startsWith(name) || x.code.value.toLowerCase.equals(name)).sortBy(_.name.value);
  }

  def sortByAirportsQty(top: Boolean, calc: Boolean, maxElements: Int) = {
    if(calc){
      countries.foreach((country) => country.airportsQty = Airports.qtyByCountry(country));
    }
    if(top) {
      countries.sortWith((c1, c2) => c1.airportsQty >= c2.airportsQty).slice(0, maxElements);
    } else {
      countries.sortWith((c1, c2) => c1.airportsQty < c2.airportsQty).slice(0, maxElements);
    }
  }

  def sortByName() = {
    countries.sortBy(_.name.value);
  }


}
